### PR TITLE
Add "steal lock" functionality next to unlock button

### PR DESCRIPTION
**Pull request by alered01**
My team has a need to be able to reserve a resource to a user while a jobs are currently running & queued up against a resource. As we do continuous testing, we aim to minimise device downtime and jobs are constantly being queued up against resources; however for development and debugging purposes - there is a need to be able to temporarily borrow devices out while maintaining the existing queue.

### CODE REVIEW

**Comment by TobiX** [comment]
Same as #144? Or at least pretty similar?

**Comment by alered01** [comment]
It sounds pretty similar; though after testing #144 on my local machine I couldn't really figure out where the reassign button was or how to use it. 
Also wasn't certain the code in #144 would behave as expected as calling unreserve() while jobs are queued up will just mean the next job will take the lock/label (unless you circumvent it - which my commit implements)

**Comment by jimklimov** [comment]
Hello, I am pleased to see the idea picked up, let the best solution win (or the two converge, if needed) ;)

As for finding the button, I never saw the Unlock one ;) In our case all the resource instances are "Reserved" by some build or person, and can only be UnReserved (so I added the stealing button there).

**Comment by jimklimov** [comment]
Also, seems there are widely different uses for the plugin out there. Many people speak about "taking labels" or specific resources by name... the use-case I've trodded most is having defined many individual resource instances with unique names that correspond to physical objects, assigned some lists of labels (not-unique) that correspond to capabilities of those devices.

So to run some test against a device, we request a label (and a count of 1 if it matters), and a particular instance is given to the job when it starts.

So in our scenario I do not see resources held by a queued job (either the job is queued because it has no device available for testing against, or a device becomes available and the job grabs it and starts "building"). Even if it indeed happens that an item is still in the queue but already with a resource, it would be some sub-second interim status that we never notice.

I would really be curious to find out if something works differently for other people. With the amount of levers in this plugin, it is quite flexible and multifaceted, and likely I only see it from one of many sides.

**Comment by alered01** [comment]
Right, I understand where you're coming from now. In that case the two functionalities are different. The steal lock button I operates on a locked resource, not a reserved one (we use groovy pipeline scripts on our jobs).

Preview of what it looks like:
chrome_5y7aKFe7Er

**Comment by jimklimov** [comment]
Thanks for confirmation. Then I think both features can be merged as addressing similar but different use-cases of the same conceptual problem :)

I wonder if there are any redundancies/conflicts in the two implementations however, and/or ways to improve mine (e.g. add locking/synchronization somewhere to ensure nothing strange happens under stress of many jobs waiting to run)?

**Comment by alered01** [comment]
In my case I've attempted to keep the changes as simple as possible. All I've done is as below:
- Pre-reserve the device to the user stealing the lock by calling reserve()
- Call unlock() on the locked device with a flag to skip the bit of code (LockStepExecution.proceed()) which will pass the lock onto the next job/request in the queue

Once the user is done with the device. When they click on the UnReserve button, regular function should continue as the call to unreserve() should call LockStepExecution.proceed() which will let the next job/request in the queue take over the lock.

**Comment by alered01** [comment]
I ran into a small problem where the lock wasn't correctly keeping when jobs failed/aborted so I've decided to store the information with the lockable resource instead so that it's tracked correctly.

**Comment by jimklimov** [comment]
I'd stand corrected. When our resources are taken by a job (legacy MultiJob usually, wrapping our old-style equivalent of pipelines) we do after all see them as "LOCKED" at some point. And so would benefit from your fix as well.

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java [hunk 1]
```diff
@@ -396,7 +396,6 @@ private synchronized void freeResources(List<String> unlockResourceNames, @Nulla
 	public synchronized void unlock(List<LockableResource> resourcesToUnLock, @Nullable Run<?, ?> build) {
 		unlock(resourcesToUnLock, build, null, false);
 	}
-
```
Comment (jimklimov) [inline src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java]: Why did this blank line between routines disappear? ;)

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java [hunk 2]
```diff
@@ -581,6 +584,17 @@ public synchronized boolean reserve(List<LockableResource> resources,
 		return true;
 	}
 
+	public synchronized boolean steal(List<LockableResource> resources,
```
Comment (jimklimov) [inline src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java]: Potential nit: would `stealResources` (as acting on many) be better? There is precedent around for both naming patterns however...

Comment (jimklimov) [reply src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java]: I'm guilty of same in #144 however. Guess having same names throughout Java and Jelly and request args was more straightforward ;)

File: src/main/java/org/jenkins/plugins/lockableresources/actions/LockableResourcesRootAction.java [hunk 3]
```diff
@@ -126,6 +126,26 @@ public void doReserve(StaplerRequest req, StaplerResponse rsp)
...
+		if (r == null) {
+			rsp.sendError(404, "Resource not found " + name);
+			return;
+		}
+
+		List<LockableResource> resources = new ArrayList<>();
+		resources.add(r);
+		String userName = getUserName();
+		if (userName != null)
+			LockableResourcesManager.get().steal(resources, userName);
```
Comment (jimklimov) [inline src/main/java/org/jenkins/plugins/lockableresources/actions/LockableResourcesRootAction.java]: Comparing to code around https://github.com/jenkinsci/lockable-resources-plugin/pull/144/files#diff-267821931388d54e1318e5a22c6a6557R146 there is less access-permissions and related sanity checking here.

Comment (jimklimov) [reply src/main/java/org/jenkins/plugins/lockableresources/actions/LockableResourcesRootAction.java]: FWIW, at https://github.com/jimklimov/lockable-resources-plugin/tree/privateBuild I track a merger of these two PRs. They seem non-conflicting in things they change logically, but there are new different routines added in the same places so automatic git merger fails to put the two together.

Comment (alered01) [reply src/main/java/org/jenkins/plugins/lockableresources/actions/LockableResourcesRootAction.java]: That was functionally copied from the reserve function right above it with one function call alteration (.reserve -> .steal). That said I'm not sure whether it's more appropriate to do .checkPermission(RESERVE) or .checkPermission(UNLOCK) for that function. Any input would be welcome

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java (hunk 2 above)
Comment (alered01) [reply src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java]: I stuck to the layout presented to me assuming I started writing functionality for the button (while copying what the reserve button does); as such that was how all the functions ended up being named.

File: src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java (hunk 1 above)
Comment (alered01) [reply src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java]: Definitely a mistake on my end; missed it on my refactor

**Comment by jimklimov** [comment]
Tested the build of https://github.com/jimklimov/lockable-resources-plugin/tree/privateBuild with this PR and mine applied, seems to work (the resource previously "LOCKED" by a job becomes "RESERVED" by me after "stealing").

Integration-wise, if both PRs are to be merged, it makes sense to settle on the name for the button (leading to different actions depending on initial state) so users are less confused. And I'd wager somebody would find "stealing", "snatching" or "grabbing" a bit too rough... so settled on "Reassign" for my PR ;)

**Comment by alered01** [comment]
Feel free to rename/suggest names for the button if it feels appropriate, I'm only after the functionality in the end. As long as the naming explains its function, I'm not too overly fussed about what it's called.

**Comment by jimklimov** [comment]
FYI : I applied your commits to my branch in #144 so the change (and now both of them together) remains usable after a large reformatting happened in project history. So I think this PR can be closed, to reduce confusion and duplicate work to bring it into master-compatible shape again :)

**Pull request closed without merging**